AST subtypes

cdisselkoen opened the pull request:
This is a first effort to try to address #31.  Comments or code review very welcome.

There are a number of questions or ambiguities, which I've called out in code comments for now (and can remove once the questions are resolved).  For instance, with AST subtypes, I think it makes sense to remove (or at least deprecate) the generic `ctx.named_const()`, `ctx.fresh_const()`, etc methods in favor of the typed variants `ctx.named_bool_const()`, `ctx.named_int_const`, etc; but I'm not sure how important the generic variants are or how breaking that would be.

The first commit in this PR exposes a few additional functions in the high-level API that were previously only in `z3-sys`.  In particular, this means the `SortKind` enum is now a public part of the `z3` crate.  I'm assuming that's OK, but I wanted to call it out specifically just in case.  Also, if that's enough of an unrelated change to be a separate PR, I'm happy to pull it out separately.

All current tests are passing (except `test_solve_simple_semver_example` which was already failing on `master` (#2?)), but arguably many more tests would be needed to ensure that all these method types are actually correct.  I wanted to get some feedback first though before writing a bunch more tests.

For crate users, these are clearly breaking changes for the API, and would need to be semver'd appropriately if this PR is accepted.

If this PR is going to be accepted, I'll squash before it actually gets merged, in order to keep the git history clean.

Fixes #31.

waywardmonkeys commented:
Does the `SortKind` need to be exposed outside of `z3` or just used by `z3`?

waywardmonkeys commented on z3/src/lib.rs:
```diff
@@ -55,19 +56,12 @@ pub struct Symbol<'ctx> {
 }
 
-/// Kind of [`Ast`](struct.Ast.html) used to represent types.
+/// Sorts represent the various 'types' of [`Ast`s](struct.Ast.html).
```
This `struct.Ast.html` probably should become `trait.Ast.html`, right?

waywardmonkeys replied:
You got that in a followup commit already.

waywardmonkeys commented:
Having an `Into` / `From` impl to convert from the various AST types to the raw `Z3_ast` might be prettier than some of the current code. Just a thought ...

waywardmonkeys commented:
This is a good start ... there are some other nits that I have, like perhaps we can have `z3::ast::Bool` instead of `z3::BoolAst` ... and clean up the naming that way. But would you mind force pushing to this PR a cleaned up set of commits and perhaps we'll just land it and we can iterate on it from there?

I've had a terrible headache for the last couple of days, so I'm going to sign off for now and get some rest.

cdisselkoen commented:
> Does the `SortKind` need to be exposed outside of `z3` or just used by `z3`?

I had added a method `get_kind()` on `Sort` which returns `SortKind`, but I can remove it.  `z3` can just use `z3_sys`'s `Z3_get_sort_kind` when it needs the `SortKind`.  This way `SortKind` can stay internal.

cdisselkoen commented:
Changelog for that force-push:

- Removed `Sort::get_kind()` and no longer expose `z3_sys::SortKind`
- Renamed `BoolAst` to `ast::Bool`, etc. `ast` is now a `pub mod` on `z3`.
- Added `From` impls to convert the various AST types to `Z3_ast`; these imply the corresponding `Into` impls
- Squashed into a single commit

cdisselkoen commented:
Agh, I apologize.  I'm new enough to Rust that I didn't realize that `cargo test` wouldn't necessarily run all tests before quitting; I need `cargo test --no-fail-fast`.  There are a bunch more failing tests, which I will fix before this is merged.

cdisselkoen commented:
Force-pushed again.  Added some more `From` impls.  Only two doctests failing now; one I think we should just accept as a regression for the moment (the `enumeration_sort` test fails with `not yet implemented: Dynamic with kind Datatype`), and the other has me stumped (the `ast::BV::from_int()` test fails due to `as_i64()`  returning `None` on a `BV`).

cdisselkoen commented:
Fixed `ite`, which should be `(Bool, T, T) -> T` rather than `(Bool, T, T) -> Bool`

waywardmonkeys commented on z3/src/solver.rs:
```diff
@@ -205,9 +203,12 @@ impl<'ctx> Solver<'ctx> {
-    pub fn get_proof(&self) -> Ast<'ctx> {
+    //
+    // TODO: what kind of AST does this actually return? If it's consistent, we can just
+    // make that the return type.
```
In the Python bindings, this is labelled as an `ExprRef`, where the inheritance hierarchy is `AstRef` -> `ExprRef` -> (`BoolRef`, etc...).

But could we use an `-> impl Ast<'ctx>` here instead of `ast::Dynamic`?

cdisselkoen replied:
`-> impl Ast<'ctx>` seems to work in my testing.  In my simple example, the returned `Ast` had the sort `SortKind::Unknown`.

cdisselkoen replied:
Of course, we still have to use some concrete constructor in the method body.  In my testing, I was still calling `ast::Dynamic::new()` to construct the return value, so it's not terribly surprising I guess that the `-> impl Ast` worked as a declaration for the return type.

waywardmonkeys replied:
In Python, they use `_to_expr_ref`:

```
def _to_expr_ref(a, ctx):
    if isinstance(a, Pattern):
        return PatternRef(a, ctx)
    ctx_ref = ctx.ref()
    k = Z3_get_ast_kind(ctx_ref, a)
    if k == Z3_QUANTIFIER_AST:
        return QuantifierRef(a, ctx)
    sk = Z3_get_sort_kind(ctx_ref, Z3_get_sort(ctx_ref, a))
    if sk == Z3_BOOL_SORT:
        return BoolRef(a, ctx)
    if sk == Z3_INT_SORT:
        if k == Z3_NUMERAL_AST:
            return IntNumRef(a, ctx)
        return ArithRef(a, ctx)
    if sk == Z3_REAL_SORT:
        if k == Z3_NUMERAL_AST:
            return RatNumRef(a, ctx)
        if _is_algebraic(ctx, a):
            return AlgebraicNumRef(a, ctx)
        return ArithRef(a, ctx)
    if sk == Z3_BV_SORT:
        if k == Z3_NUMERAL_AST:
            return BitVecNumRef(a, ctx)
        else:
            return BitVecRef(a, ctx)
    if sk == Z3_ARRAY_SORT:
        return ArrayRef(a, ctx)
    if sk == Z3_DATATYPE_SORT:
        return DatatypeRef(a, ctx)
    if sk == Z3_FLOATING_POINT_SORT:
        if k == Z3_APP_AST and _is_numeral(ctx, a):
            return FPNumRef(a, ctx)
        else:
            return FPRef(a, ctx)
    if sk == Z3_FINITE_DOMAIN_SORT:
        if k == Z3_NUMERAL_AST:
            return FiniteDomainNumRef(a, ctx)
        else:
            return FiniteDomainRef(a, ctx)
    if sk == Z3_ROUNDING_MODE_SORT:
        return FPRMRef(a, ctx)
    if sk == Z3_SEQ_SORT:
        return SeqRef(a, ctx)
    if sk == Z3_RE_SORT:
        return ReRef(a, ctx)
    return ExprRef(a, ctx)
```

But for the proof to be generated, the `check` / `check_assumptions` must have failed and proof generation must be enabled on the context.  In C, you'd do that via:

```
    Z3_set_param_value(cfg, "proof", "true");
```

cdisselkoen replied:
Right, which is called by `Config::set_proof_generation()` in Rust if I'm not mistaken.

So I guess this really is an `Ast` of no specialized type then.  We could create a minimal `Ast` subtype for this, we could even call it `Proof` if you want.  Is that the best way forwards?

waywardmonkeys replied:
Let's leave it as is for now, I guess and then we can iterate on it and experiment and think.

cdisselkoen replied:
I interpret "as is" as "as `-> impl Ast`" for now, so that's in the last force-push. I think `-> impl Ast` is probably a better analogue to the Python `ExprRef` type than `Dynamic` is, anyway.

waywardmonkeys commented:
Also, thanks for doing this! The work is definitely appreciated and a great step forwards.

I suspect that we're almost at a place where we can merge it and then work on improving it from there...

waywardmonkeys commented:
Apart from `Dynamic` ... is there anything blocking merging this?

cdisselkoen commented:
What specifically about `Dynamic` do you want addressed?

I'll make a few more comments inline about some other small issues.

cdisselkoen commented on z3/src/ast.rs:
```diff
@@ -381,75 +381,975 @@ impl<'ctx> Ast<'ctx> {
+    binop!(ge, Z3_mk_ge, Bool<'ctx>);
+    // Z3 does support mixing ints and reals in add(), sub(), mul(), div(), and power()
+    //   (but not rem(), modulo(), lt(), le(), gt(), or ge()).
+    // TODO: we could consider expressing this by having a Numeric trait with these methods.
```
There's this question about mixing ints and reals in certain methods (which Z3 supports but the current draft of this PR does not)

cdisselkoen replied:
(see the rest of this comment as well, continues to line 527)

cdisselkoen commented on z3/src/context.rs:
```diff
@@ -94,6 +94,9 @@ impl Context {
 
+/* Deprecate in favor of the typed variants below?
+    Or even better, the {Bool,Int,etc}Ast::new_const() and ::fresh_const() methods?
+
```
this question about deprecating the untyped methods on `Context`

waywardmonkeys replied:
Leave and then think about it later.

cdisselkoen replied:
If we leave these in, should they just return `Dynamic` for now?

waywardmonkeys replied:
Oh, right, these are commented out. Removing them is fine then.

cdisselkoen replied:
Removed in the latest push

cdisselkoen commented on z3/src/context.rs:
```diff
@@ -105,69 +108,71 @@ impl Context {
+    // Deprecate from_u64() and from_i64() in favor of
+    // ast::Int::from_u64()/from_i64() and ast::BV::from_u64()/from_i64()
+    pub fn from_u64(&self, u: u64) -> ast::Int {
+        ast::Int::from_u64(self, u)
```
this is another deprecation question

waywardmonkeys replied:
Deprecate / remove.

cdisselkoen replied:
These turn out to be used a bunch even in the existing tests, so I'll just deprecate.

cdisselkoen commented on z3/src/sort.rs:
```diff
@@ -212,6 +212,8 @@ impl<'ctx> Sort<'ctx> {
         })
     }
+*/
+
```
I commented out the `Sort::from_u64(&self, u: u64)` and `Sort::from_i64(&self, i: i64)` methods; if you're okay with that, we can just remove them before merging rather than leaving them commented out

waywardmonkeys replied:
Okay to remove, I think.

cdisselkoen commented:
and then my final comment is that we should probably add more doctests for most of the methods.  For instance, I accidentally had the wrong type for `ite` at first until I fixed it earlier today, as noted above in the thread.  But we can probably merge now and add tests later if you're comfortable with that.

waywardmonkeys commented:
Okay, so if you push with the methods on `Sort` removed, I'll merge and then we can iterate. Doctests can definitely come after the merge in this case.

cdisselkoen commented:
Pushed, and I think that's all of the issues that need to be addressed before merge.  I'm 👍 to merge

waywardmonkeys commented:
Thanks!
